[STORM-1270]  [STORM-1274] port drpc&LocalDRPC  to java

hustfxj opened the pull request:
1. Port "org.apache.storm.daemon.drpc" to java at "org.apache.storm.daemon.DrpcServer".
2. Port "org.apache.storm.LocalDRPC" to java at "org.apache.storm.LocalDRPC".
3. I keep the "webapp" of drpc.clj , as it's hard to translate and we will move ring to other REST framework.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class DrpcServer implements DistributedRPC.Iface, DistributedRPCInvocations.Iface, Shutdownable {
```
you can rename it to DRPCServer for consistency.

abhishekagarwal87 replied:
You can separate out the implementation of DistributedRPC.Iface, DistributedRPCInvocations.Iface, Shutdownable to a class DRPCService. This service class will only contain the business logic and server logic can be inside DRPCServer. Also AutoCloseable should be the interface than Shutdownable.

hustfxj replied:
I think we can rename DRPCServer to DRPC later

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+        });
+        LOG.info("Starting Distributed RPC servers...");
+
+        LOG.info("Starting Distributed RPC servers...");
```
Duplicate log statement

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+    }
+
+    public static void main(String[] args) throws Exception {
+
```
There are extra lines in some places which you can remove.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+        for (PreparableReporter reporter : reporters) {
+            reporter.prepare(new MetricRegistry(), conf);
+            reporter.start();
+            LOG.info("Started statistics report plugin...");
```
This will be logged multiple times or you can add the reporter class

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+            });
+        }
+
+        // To be replaced by Common.StartMetricsReporters
```
you can add a TODO so that it is not missed

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+        Semaphore sem = new Semaphore(0);
+
+        DRPCRequest req = new DRPCRequest(funcArgs, strid);
+        this.idtoStart.put(strid, Time.currentTimeSecs());
```
minor. can you rename idtoStart to idToStart and similarly other idto* variables.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+
+        Object result = this.idtoResult.get(strid);
+
+        LOG.info("Returning for DRPC request for " + functionName + " " + funcArgs + " at " + (System.currentTimeMillis()));
```
this should be LOG.debug

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+            try {
+                aznClass = Class.forName(klassname);
+                authorizer = (IAuthorizer) aznClass.newInstance();
+                if (authorizer != null) {
```
newInstance shouldn't return null

revans2 replied:
This is still an issue to be fixed.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+    private ConcurrentHashMap<String, DRPCRequest> idtoRequest = new ConcurrentHashMap<String, DRPCRequest>();
+    private ConcurrentHashMap<String, ConcurrentLinkedQueue<DRPCRequest>> requestQueues = new ConcurrentHashMap<String, ConcurrentLinkedQueue<DRPCRequest>>();
+
+    private final Meter meterHttpRequests = new MetricRegistry().meter("drpc:num-execute-http-requests");
```
this variable is not used

hustfxj replied:
meterHttpRequests is not used as drpc.clj has meter for httpRequests. But we can use this meter after we remove drpc.clj

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+            public void run() {
+                if (handlerServer != null) {
+                    handlerServer.stop();
+                } else {
```
shouldn't it try to stop both of them

revans2 replied:
Yes it should the original code was more of

```
if (handlerServer != null) handlerServer.stop();
invokeServer.stop();
```

hustfxj replied:
sorry. I make mistake

revans2 commented on examples/storm-starter/src/jvm/org/apache/storm/starter/ManualDRPC.java:
```diff
@@ -30,39 +30,37 @@
+        System.out.println(drpc.execute("exclamation", "aaa"));
+        System.out.println(drpc.execute("exclamation", "bbb"));
+    }
 }
```
Please don't change the spacing/formatting for files you didn't touch otherwise.

hustfxj replied:
I will pay attention to this，Thank you for reminding me

revans2 commented on storm-core/src/clj/org/apache/storm/daemon/drpc.clj:
```diff
@@ -210,67 +63,19 @@
-      (start-metrics-reporters conf)
-      (when handler-server
-        (.serve handler-server)))))
+        (let [app (-> (webapp drpc-server http-creds-handler)
```
Can we please have a follow on JIRA to port the webapp portion over to java too?  Until it is done I don't consider DRPC moved to java.

hustfxj replied:
ok

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+
+        DRPCRequest req = new DRPCRequest(funcArgs, strid);
+        this.idtoStart.put(strid, Time.currentTimeSecs());
+        this.idtoSem.put(strid, sem);
```
To make code more clean, can we have methods like below
```
String registerRequest(functionName, functionArgs) - Creates a new id with the function name
awaitOnResult(id) - blocks on the semaphore associated with the id
notifyOnResult(id) - releases the semaphore associated with the id
updateResult(id, Object result) - update the result 
removeRequest(id) - Cleanup the request
```
The above four methods can go into a class named DRPCState/DRPCRequestRegistry. what do you think?

abhishekagarwal87 replied:
All the semaphore/queue related logic will sit together in one class and not dispersed across.

hustfxj replied:
@abhishekagarwal87  DRPC have already result fetchRequest execute failRequest， and these are override. I don't want to add some other methods.

abhishekagarwal87 replied:
You can add them in the InternalRequest class suggested by Robert.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.MetricRegistry;
+import com.sun.net.httpserver.HttpsServer;
+import com.sun.org.apache.bcel.internal.generic.ARRAYLENGTH;
```
We should not be using com.sun internal classes.  We didn't use them before we should not use them now.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+    private ConcurrentHashMap<String, Integer> idtoStart = new ConcurrentHashMap<String, Integer>();
+    private ConcurrentHashMap<String, String> idtoFunction = new ConcurrentHashMap<String, String>();
+    private ConcurrentHashMap<String, DRPCRequest> idtoRequest = new ConcurrentHashMap<String, DRPCRequest>();
+    private ConcurrentHashMap<String, ConcurrentLinkedQueue<DRPCRequest>> requestQueues = new ConcurrentHashMap<String, ConcurrentLinkedQueue<DRPCRequest>>();
```
This is a perfect place to actually improve things to make them more object oriented.  The key to almost all of these maps is the same thing.  Why not remove all of the special ordering that we have to do on setup and cleanup.

```
private static class InternalRequest {
    public final Semaphore sem;
    public final long startTimeSecs;
    public final string function;
    public final DRPCRequest request;
    public volatile Object result;

    public InternalRequest(String function, DRPCRequest request) {
        sem = new Semaphore(0);
        startTimeSecs = Time.currentTimeSecs();
        this.function = function;
        this.request = request;
    }
}

private ConcurrentHashMap<String, InternalRequest> outstandingRequests = new ConcurrentHashMap<>();
```

hustfxj replied:
perfect   thank you

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+    private final Meter meterResultCalls = new MetricRegistry().meter("drpc:num-result-calls");
+    private final Meter meterFailRequestCalls = new MetricRegistry().meter("drpc:num-failRequest-calls");
+    private final Meter meterFetchRequestCalls = new MetricRegistry().meter("drpc:num-fetchRequest-calls");
+    private final Meter meterShutdownCalls = new MetricRegistry().meter("drpc:num-shutdown-calls");
```
We cannot just create a new MetricRegistry each time.  The metrics will be lost.  In the clojure code there is a default registry that is used by all of the metrics.  We need a similar one everywhere for all of our code to use (including the clojure code)

hustfxj replied:
Can we  follow up a JIRA to port a default registry over to java too? I will do it.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+
+    public Servlet getHttpServlet() {
+        return httpServlet;
+    }
```
This is never used and should be removed.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+
+    public void setHttpCredsHandler(IHttpCredentialsPlugin httpCredsHandler) {
+        this.httpCredsHandler = httpCredsHandler;
+    }
```
the set and get of HttpCredsHandler is never used.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+        return conf;
+    }
+
+    public static void main(String[] args) throws Exception {
```
This main does not work if we are using HTTP.  The servlet is not setup.  We should probably remove this until it actually works.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+        return authorizer;
+    }
+
+    public Map getConf() {
```
This is never used.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+        return timeoutCheckSecs;
+    }
+
+    public void launchServer(boolean isLocal, Map conf) throws Exception {
```
Why did we combine service-handler and launch-server! into a single launchServer method?  The only thing that was missing in local mode that this adds is the clearThread to time out missed tuples.  Did you run into issues with testing when this was not included?

revans2 replied:
Never mind the old code launched the clear-thread in service-handler.  To me it feels cleaner to have the Constructor act like service-handler and launchServer to start up the thrift servers and HTTP Server.

revans2 replied:
If we move the initClearThread to be called from the constructor, and have the constructor take a conf we can then remove the need for any parameters top be passed in here, and simply have localmode not call this method.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+    @Override
+    public String execute(String functionName, String funcArgs) throws DRPCExecutionException, AuthorizationException, org.apache.thrift.TException {
+        meterExecuteCalls.mark();
+        LOG.debug("Received DRPC request for {} {} at {} ", functionName, funcArgs, System.currentTimeMillis());
```
minor nit. There were parens around the funcArgs

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+        meterExecuteCalls.mark();
+        LOG.debug("Received DRPC request for {} {} at {} ", functionName, funcArgs, System.currentTimeMillis());
+        Map<String, String> map = new HashMap<>();
+        map.put(DRPCAuthorizerBase.FUNCTION_NAME, functionName);
```
Does Guava have a way to create this map cleaner?

abhishekagarwal87 replied:
ImmutableMap.of(DRPCAuthorizerBase.FUNCTION_NAME, functionName)

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+        int newid = idinc % maxvalue;
+        if (idinc != newid) {
+            this.ctr.compareAndSet(idinc, newid);
+        }
```
This does not actually do the same thing.  what happens if the compareAndSet fails?  If you want it the same we have to loop around and try again until it succeeds.

```
int newid = 0;
do {
  int orig = ctr.get();
  newid = (orig + 1) % 1000000000;
} while (!ctr.compareAndSet(orig, newid));
String strid = String.valueOf(newid);
```

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+        if (result == null) {
+            throw new DRPCExecutionException("Request timed out");
+        }
+        return String.valueOf(result);
```
Just cast result to a String, if it is not a String we have an internal error that we need to fix.

Or better yet, in that class I was suggesting we create.  Instead of having an Object for the result lets split it into an Exception and a String result.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+        ConcurrentLinkedQueue<DRPCRequest> reqQueue = requestQueues.get(function);
+        if (reqQueue == null) {
+            reqQueue = new ConcurrentLinkedQueue<DRPCRequest>();
+            requestQueues.put(function, reqQueue);
```
This should be a putIfAbsent.

```
protected ConcurrentLinkedQueue<DRPCRequest> acquireQueue(String function) {
  ConcurrentLinkedQueue<DRPCRequest> reqQueue = requestQueues.get(function);
  if (reqQueue == null) {
    reqQueue = new ConcurrentLinkedQueue<>();
    ConcurrentLinkedQueue<DRPCRequest> old = requestQueues.putIfAbsent(function, reqQueue);
    if (old != null) {
      reqQueue = old;
    }
  }
  return reqQueue;
}
```

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,390 @@
+        checkAuthorization(aclHandler, mapping, operation, ReqContext.context());
+    }
+
+    // TO be replaced by Common.mkAuthorizationHandler
```
Then lets create the file and just put it in there now, so the next person does not have to do it later.

hustfxj replied:
common.clj maybe submit soon

revans2 replied:
OK

revans2 commented:
Done with my first pass through the code.

hustfxj commented:
@revans2 @abhishekagarwal87 Than you ver much. I are addressing your comments. But I may not understand what you mean about launchServer method, although I update the method. Can you help me review again?  I want to follow up a JIRA to port the webapp portion over to java about webapp,
And I want to follow up a JIRA to port a default registry over to java for all of our code to use;

revans2 commented on storm-core/src/jvm/org/apache/storm/LocalDRPC.java:
```diff
@@ -0,0 +1,80 @@
+    private static final Logger LOG = Logger.getLogger(LocalDRPC.class);
+
+    private DrpcServer handler = new DrpcServer();
+    private Thread thread;
```
This is never used and should be removed.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,383 @@
+
+    private IAuthorizer authorizer;
+
+    //TODO: To be removed after porting drpc.clj
```
Just delete the comment, the JIRA you filed should cover this.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,383 @@
+    private ConcurrentHashMap<String, InternalRequest> outstandingRequests = new ConcurrentHashMap<>();
+
+
+    //TODO: to be replaced by a common registry
```
I don't want to merge in code with broken functionality.  Lets pull in the metrics changes you made first before we pull this JIRA in.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,383 @@
+
+    }
+
+    //TODO: to be removed
```
Please remove this comment.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,383 @@
+    private final static Meter meterShutdownCalls = new MetricRegistry().meter("drpc:num-shutdown-calls");
+    
+    public DrpcServer() {
+
```
In the original code clear-thread was started in the service-handler (which is the equivalent of the constructor).  I would like to see it started here too, but that will require the constructor to take the Config.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,383 @@
+            public Object call() throws Exception {
+                for (Map.Entry<String, InternalRequest> e : outstandingRequests.entrySet()) {
+                    InternalRequest internalRequest = e.getValue();
+                    if (Time.deltaSecs(internalRequest.startTimeSecs) > Utils.getInt(conf.get(Config.DRPC_REQUEST_TIMEOUT_SECS), 0)) {
```
There was no default of 0 in the original code and we should probably not have one here either.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,383 @@
+                            sem.release();
+                        }
+                        cleanup(id);
+                        LOG.info("Clear request " + id);
```
This log did not exist in the original code.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,383 @@
+            }
+        }).start();
+
+        //TODO: To be replaced by Common.StartMetricsReporters
```
This should probably be moved into #1171 and just do it there.

revans2 replied:
That way it can use the same MetricsRegistry as the others, which is the key to make this all work.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,383 @@
+            return String.valueOf(result);
+        }catch (Exception e){
+            throw new DRPCExecutionException(e.getMessage());
+        }
```
Lets replace this the entire try/catch with.
```
return (String) result;
```

If we ever have anything but a DRPCExecutorionException or a String in there we have an internal error that we need to fix.

revans2 commented on storm-core/test/clj/org/apache/storm/security/auth/auth_test.clj:
```diff
@@ -27,6 +27,8 @@
   (:import [java.net InetAddress])
   (:import [org.apache.storm Config])
+  (:import [org.mockito Mockito])
+  (:import [org.mockito.exceptions.base MockitoAssertionError])
```
Are these needed?  If not please revert.  If so how did this work before?

revans2 commented:
Done with another pass

hustfxj commented:
@revans2  Thank you.I won't wait updating the metric codes util the #1171  about the metrics is merged master.

hustfxj commented:
@revans2 @abhishekagarwal87 could you take a look?

abhishekagarwal87 commented on storm-core/src/clj/org/apache/storm/daemon/supervisor.clj:
```diff
@@ -1297,7 +1297,7 @@
       (.readBlobTo blob-store (ConfigUtils/masterStormConfKey storm-id) (FileOutputStream. (ConfigUtils/supervisorStormConfPath tmproot)) nil)
       (finally
         (.shutdown blob-store)))
-    (FileUtils/moveDirectory (File. tmproot) (File. stormroot))
```
this file can be reverted.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,356 @@
+        return reqQueue;
+    }
+
+    private void checkAuthorization(IAuthorizer aclHandler, Map mapping, String operation, ReqContext reqContext) throws AuthorizationException {
```
This is not called in this class and has been marked private. So it can be removed I believe.

revans2 replied:
It is called, on line 337
```
 private void checkAuthorization(IAuthorizer aclHandler, Map mapping, String operation) throws AuthorizationException {
    checkAuthorization(aclHandler, mapping, operation, ReqContext.context());
}
```

abhishekagarwal87 replied:
@revans2 I meant there is no need of overloading function as only one signature is being called. But its fine to merge still.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,356 @@
+        if (internalRequest != null) {
+            Map<String, String> map = new HashMap<>();
+            map.put(DRPCAuthorizerBase.FUNCTION_NAME, internalRequest.function);
+            checkAuthorization(authorizer, map, "failRequest");
```
It would be good to have operation names declared as public static variables.

abhishekagarwal87 commented:
Thanks @hustfxj . Looks good to me.

revans2 commented:
@hustfxj looks good to me I am +1 on this.  @abhishekagarwal87 are you OK with merging this in as is.  The comments you had looked relatively minor.

abhishekagarwal87 commented:
Yes. +1

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java:
```diff
@@ -0,0 +1,356 @@
+        StormMetricsRegistry.startMetricsReporters(conf);
+
+        if (handlerServer != null)
+            handlerServer.serve();
```
Sorry I missed this before.  In the previous code this was the very last thing that happened on the main thread, because this is a blocking call.  With this happening before initHttp the HTTP server is never initialized even when it is configured.

revans2 replied:
Looks like you can just switch the order of doing the init and put initHttp() first in launchServer, but I am getting some problems with using HTTP with the following.

```
$ cat ./conf/storm.yaml
drpc.http.port: 9090
drpc.servers:
    - "localhost"
$ ./bin/storm jar ./examples/storm-starter/storm-starter-topologies-2.0.0-SNAPSHOT.jar org.apache.storm.starter.BasicDRPCTopology drpc-test
... wait for to come all the way up
$ curl http://localhost:9090/exclamation/hi
... Error ...
```

from the logs I see
```
2016-03-08 15:17:49.528 o.a.s.s.o.e.j.s.ServletHandler [WARN] /exclamation/hi
java.lang.NullPointerException: Handler returned nil
        at org.apache.storm.shade.ring.util.servlet$make_service_method$fn__7738.invoke(servlet.clj:128)
        at org.apache.storm.shade.ring.util.servlet$servlet$fn__7742.invoke(servlet.clj:136)
        at org.apache.storm.shade.ring.util.servlet.proxy$javax.servlet.http.HttpServlet$ff19274a.service(Unknown Source)
        at org.apache.storm.shade.org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:654)
        at org.apache.storm.shade.org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1320)
        at org.apache.storm.logging.filters.AccessLoggingFilter.handle(AccessLoggingFilter.java:47)
        at org.apache.storm.logging.filters.AccessLoggingFilter.doFilter(AccessLoggingFilter.java:39)
        at org.apache.storm.shade.org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1291)
        at org.apache.storm.shade.org.eclipse.jetty.servlets.CrossOriginFilter.handle(CrossOriginFilter.java:247)
```

revans2 replied:
OK So the issue with http not working is that the handler is never set for the http server to have access to.

hustfxj replied:
@revans2  I can see the same errors when I run Drpc of master branch.  Later I find we maybe misunderstand it. As follows:
$ curl http://localhost:3774/drpc/exclamation/hi
hi!
It's ok. So the URL should like ""/drpc/:func/*"

hustfxj replied:
@revans2  I have adjust the order of initHttp(). Thank you again.

revans2 replied:
Crap you are right, I totally messed up the URL.  Great work!

revans2 commented:
+1
